File: yizhi/111/resources/com/dev/xalu/modules/NoSlow.java
package com.dev.xalu.modules;

import com.dev.xalu.XALUAddon;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import meteordevelopment.meteorclient.events.packets.PacketEvent;
import meteordevelopment.meteorclient.events.world.TickEvent;
import meteordevelopment.meteorclient.settings.BoolSetting;
import meteordevelopment.meteorclient.settings.DoubleSetting;
import meteordevelopment.meteorclient.settings.Setting;
import meteordevelopment.meteorclient.settings.SettingGroup;
import meteordevelopment.meteorclient.systems.modules.Module;
import meteordevelopment.orbit.EventHandler;
import net.minecraft.class_1268;
import net.minecraft.class_1799;
import net.minecraft.class_1802;
import net.minecraft.class_2338;
import net.minecraft.class_2350;
import net.minecraft.class_238;
import net.minecraft.class_2560;
import net.minecraft.class_2813;
import net.minecraft.class_2828;
import net.minecraft.class_2846;
import net.minecraft.class_2848;
import net.minecraft.class_2868;
import net.minecraft.class_2886;
import net.minecraft.class_304;
import net.minecraft.class_3532;
import net.minecraft.class_3675;
import net.minecraft.class_408;
import net.minecraft.class_418;
import net.minecraft.class_498;
import net.minecraft.class_5134;
import net.minecraft.class_9334;

/* JADX INFO: loaded from: XALU Beta Version-1.2.1.jar:com/dev/xalu/modules/NoSlow.class */
public class NoSlow extends Module {
    public static NoSlow INSTANCE;
    private final SettingGroup sgGeneral;
    private final SettingGroup sgBlocks;
    private final Setting<Boolean> strict;
    private final Setting<Boolean> airStrict;
    private final Setting<Boolean> grim;
    private final Setting<Boolean> grimNew;
    private final Setting<Boolean> strafeFix;
    private final Setting<Boolean> inventoryMove;
    private final Setting<Boolean> arrowMove;
    private final Setting<Boolean> items;
    private final Setting<Boolean> sneak;
    private final Setting<Boolean> crawl;
    private final Setting<Boolean> shields;
    private final Setting<Boolean> webs;
    private final Setting<Boolean> berryBush;
    private final Setting<Double> webSpeed;
    private final Setting<Boolean> soulSand;
    private final Setting<Boolean> honeyBlock;
    private final Setting<Boolean> slimeBlock;
    private boolean sneaking;

    public NoSlow() {
        super(XALUAddon.CATEGORY, "NoSlow", "Prevents slowing down when using items");
        this.sgGeneral = this.settings.getDefaultGroup();
        this.sgBlocks = this.settings.createGroup("Blocks");
        this.strict = this.sgGeneral.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("strict")).description("Strict mode for some anti-cheats")).defaultValue(false)).build());
        this.airStrict = this.sgGeneral.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("air-strict")).description("Air strict mode")).defaultValue(false)).build());
        this.grim = this.sgGeneral.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("grim")).description("Grim anti-cheat bypass")).defaultValue(false)).build());
        this.grimNew = this.sgGeneral.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("grim-v3")).description("Grim v3 bypass")).defaultValue(false)).build());
        this.strafeFix = this.sgGeneral.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("strafe-fix")).description("Fix strafe movement")).defaultValue(false)).build());
        this.inventoryMove = this.sgGeneral.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("inventory-move")).description("Allow moving in inventory screens")).defaultValue(true)).build());
        SettingGroup settingGroup = this.sgGeneral;
        BoolSetting.Builder builder = (BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("arrow-move")).description("Allow looking around with arrow keys")).defaultValue(false);
        Setting<Boolean> setting = this.inventoryMove;
        Objects.requireNonNull(setting);
        this.arrowMove = settingGroup.add(((BoolSetting.Builder) builder.visible(setting::get)).build());
        this.items = this.sgGeneral.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("items")).description("No slow when using items")).defaultValue(true)).build());
        this.sneak = this.sgGeneral.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("sneak")).description("No slow when sneaking")).defaultValue(false)).build());
        this.crawl = this.sgGeneral.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("crawl")).description("No slow when crawling")).defaultValue(false)).build());
        this.shields = this.sgGeneral.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("shields")).description("No slow when using shields")).defaultValue(true)).build());
        this.webs = this.sgBlocks.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("webs")).description("No slow in cobwebs")).defaultValue(false)).build());
        this.berryBush = this.sgBlocks.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("berry-bush")).description("No slow in berry bushes")).defaultValue(false)).build());
        this.webSpeed = this.sgBlocks.add(((DoubleSetting.Builder) ((DoubleSetting.Builder) ((DoubleSetting.Builder) new DoubleSetting.Builder().name("web-multiplier")).description("Speed multiplier in webs")).defaultValue(1.0d).min(0.0d).sliderMax(1.0d).visible(() -> {
            return ((Boolean) this.webs.get()).booleanValue() || ((Boolean) this.berryBush.get()).booleanValue();
        })).build());
        this.soulSand = this.sgBlocks.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("soul-sand")).description("No slow on soul sand")).defaultValue(false)).build());
        this.honeyBlock = this.sgBlocks.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("honey-block")).description("No slow on honey blocks")).defaultValue(false)).build());
        this.slimeBlock = this.sgBlocks.add(((BoolSetting.Builder) ((BoolSetting.Builder) ((BoolSetting.Builder) new BoolSetting.Builder().name("slime-block")).description("No slow on slime blocks")).defaultValue(false)).build());
        this.sneaking = false;
        INSTANCE = this;
    }

    public void onDeactivate() {
        if (((Boolean) this.airStrict.get()).booleanValue() && this.sneaking && this.mc.field_1724 != null) {
            this.mc.method_1562().method_52787(new class_2848(this.mc.field_1724, class_2848.class_2849.field_12984));
        }
        this.sneaking = false;
    }

    @EventHandler
    private void onTick(TickEvent.Pre event) {
        if (this.mc.field_1724 == null || this.mc.field_1687 == null) {
            return;
        }
        if (((Boolean) this.grim.get()).booleanValue() && this.mc.field_1724.method_6115() && !this.mc.field_1724.method_5715() && ((Boolean) this.items.get()).booleanValue()) {
            if (this.mc.field_1724.method_6058() == class_1268.field_5810 && checkStack(this.mc.field_1724.method_6047())) {
                sendInteractPacket(class_1268.field_5808);
            } else if (checkStack(this.mc.field_1724.method_6079())) {
                sendInteractPacket(class_1268.field_5810);
            }
        }
        if (((Boolean) this.airStrict.get()).booleanValue() && !this.mc.field_1724.method_6115()) {
            this.sneaking = false;
            this.mc.method_1562().method_52787(new class_2848(this.mc.field_1724, class_2848.class_2849.field_12984));
        }
        if (((Boolean) this.inventoryMove.get()).booleanValue() && checkScreen()) {
            long handle = this.mc.method_22683().method_4490();
            class_304[] keys = {this.mc.field_1690.field_1903, this.mc.field_1690.field_1894, this.mc.field_1690.field_1881, this.mc.field_1690.field_1849, this.mc.field_1690.field_1913};
            for (class_304 binding : keys) {
                binding.method_23481(class_3675.method_15987(handle, binding.method_1429().method_1444()));
            }
            if (((Boolean) this.arrowMove.get()).booleanValue()) {
                float yaw = this.mc.field_1724.method_36454();
                float pitch = this.mc.field_1724.method_36455();
                if (class_3675.method_15987(handle, 265)) {
                    pitch -= 3.0f;
                } else if (class_3675.method_15987(handle, 264)) {
                    pitch += 3.0f;
                } else if (class_3675.method_15987(handle, 263)) {
                    yaw -= 3.0f;
                } else if (class_3675.method_15987(handle, 262)) {
                    yaw += 3.0f;
                }
                this.mc.field_1724.method_36456(yaw);
                this.mc.field_1724.method_36457(class_3532.method_15363(pitch, -90.0f, 90.0f));
            }
        }
        if ((((Boolean) this.grim.get()).booleanValue() || ((Boolean) this.grimNew.get()).booleanValue()) && ((Boolean) this.webs.get()).booleanValue()) {
            class_238 bb = ((Boolean) this.grim.get()).booleanValue() ? this.mc.field_1724.method_5829().method_1014(1.0d) : this.mc.field_1724.method_5829();
            for (class_2338 pos : getIntersectingWebs(bb)) {
                this.mc.method_1562().method_52787(new class_2846(class_2846.class_2847.field_12973, pos, class_2350.field_11033));
            }
        }
        if (((Boolean) this.airStrict.get()).booleanValue() && !this.sneaking && checkSlowed()) {
            this.sneaking = true;
            this.mc.method_1562().method_52787(new class_2848(this.mc.field_1724, class_2848.class_2849.field_12979));
        }
        handleMovementSlowdown();
    }

    @EventHandler
    private void onPacketSend(PacketEvent.Send event) {
        if (this.mc.field_1724 == null || this.mc.field_1687 == null || this.mc.method_1542()) {
            return;
        }
        class_2828 class_2828Var = event.packet;
        if (class_2828Var instanceof class_2828) {
            class_2828 packet = class_2828Var;
            if (packet.method_36171() && ((Boolean) this.strict.get()).booleanValue() && checkSlowed()) {
                this.mc.method_1562().method_52787(new class_2868(this.mc.field_1724.method_31548().field_7545));
                return;
            }
        }
        if ((event.packet instanceof class_2813) && ((Boolean) this.strict.get()).booleanValue()) {
            if (this.mc.field_1724.method_6115()) {
                this.mc.field_1724.method_6075();
            }
            if (this.sneaking || this.mc.field_1724.method_5715()) {
                this.mc.method_1562().method_52787(new class_2848(this.mc.field_1724, class_2848.class_2849.field_12984));
            }
            if (this.mc.field_1724.method_5624()) {
                this.mc.method_1562().method_52787(new class_2848(this.mc.field_1724, class_2848.class_2849.field_12985));
            }
        }
    }

    public void handleMovementSlowdown() {
        if (this.mc.field_1724 == null) {
            return;
        }
        if ((((Boolean) this.sneak.get()).booleanValue() && this.mc.field_1724.method_5715()) || (((Boolean) this.crawl.get()).booleanValue() && this.mc.field_1724.method_20448())) {
            float f = 1.0f / ((float) this.mc.field_1724.method_45325(class_5134.field_51584));
            this.mc.field_1724.field_3913.field_3905 *= f;
            this.mc.field_1724.field_3913.field_3907 *= f;
        }
        if (checkSlowed()) {
            this.mc.field_1724.field_3913.field_3905 *= 5.0f;
            this.mc.field_1724.field_3913.field_3907 *= 5.0f;
        }
    }

    private boolean checkGrimNew() {
        return !(this.mc.field_1724.method_5715() || this.mc.field_1724.method_20448() || this.mc.field_1724.method_3144() || this.mc.field_1724.method_6014() >= 5) || (this.mc.field_1724.method_6048() > 1 && this.mc.field_1724.method_6048() % 2 != 0);
    }

    public boolean checkSlowed() {
        return (!((Boolean) this.grimNew.get()).booleanValue() || checkGrimNew()) && !this.mc.field_1724.method_3144() && !this.mc.field_1724.method_5715() && ((this.mc.field_1724.method_6115() && ((Boolean) this.items.get()).booleanValue()) || (this.mc.field_1724.method_6039() && ((Boolean) this.shields.get()).booleanValue() && !((Boolean) this.grimNew.get()).booleanValue() && !((Boolean) this.grim.get()).booleanValue()));
    }

    private boolean checkStack(class_1799 stack) {
        return (stack.method_57353().method_57832(class_9334.field_50075) || stack.method_7909() == class_1802.field_8102 || stack.method_7909() == class_1802.field_8399 || stack.method_7909() == class_1802.field_8255) ? false : true;
    }

    private void sendInteractPacket(class_1268 hand) {
        this.mc.method_1562().method_52787(new class_2886(hand, 0, this.mc.field_1724.method_36454(), this.mc.field_1724.method_36455()));
    }

    public boolean checkScreen() {
        return (this.mc.field_1755 == null || (this.mc.field_1755 instanceof class_408) || (this.mc.field_1755 instanceof class_498) || (this.mc.field_1755 instanceof class_418)) ? false : true;
    }

    public List<class_2338> getIntersectingWebs(class_238 boundingBox) {
        List<class_2338> blocks = new ArrayList<>();
        int minX = (int) Math.floor(boundingBox.field_1323);
        int maxX = (int) Math.floor(boundingBox.field_1320);
        int minY = (int) Math.floor(boundingBox.field_1322);
        int maxY = (int) Math.floor(boundingBox.field_1325);
        int minZ = (int) Math.floor(boundingBox.field_1321);
        int maxZ = (int) Math.floor(boundingBox.field_1324);
        for (int x = minX; x <= maxX; x++) {
            for (int y = minY; y <= maxY; y++) {
                for (int z = minZ; z <= maxZ; z++) {
                    class_2338 pos = new class_2338(x, y, z);
                    if (this.mc.field_1687.method_8320(pos).method_26204() instanceof class_2560) {
                        blocks.add(pos);
                    }
                }
            }
        }
        return blocks;
    }

    public boolean noSlow() {
        return isActive() && checkSlowed();
    }

    public boolean getStrafeFix() {
        return isActive() && ((Boolean) this.strafeFix.get()).booleanValue();
    }
}

File: src/main/kotlin/com/zychen027/MeteorPlusPlus/modules/NoSlow.kt
package com.zychen027.meteorplusplus.modules

import com.zychen027.meteorplusplus.MeteorPlusPlusAddon
import meteordevelopment.meteorclient.events.packets.PacketEvent
import meteordevelopment.meteorclient.events.world.TickEvent
import meteordevelopment.meteorclient.settings.BoolSetting
import meteordevelopment.meteorclient.settings.DoubleSetting
import meteordevelopment.meteorclient.systems.modules.Module
import meteordevelopment.orbit.EventHandler
import net.minecraft.block.CobwebBlock
import net.minecraft.entity.effect.StatusEffects
import net.minecraft.item.Items
import net.minecraft.network.packet.c2s.play.ClientCommandC2SPacket
import net.minecraft.network.packet.c2s.play.PlayerInteractItemC2SPacket
import net.minecraft.network.packet.c2s.play.PlayerMoveC2SPacket
import net.minecraft.util.Hand
import net.minecraft.util.math.Box

/**
 * NoSlow (无减速) - 完整移植自 XALU
 * 防止使用物品时减速，包含 Grim 反作弊绕过
 * 
 * 反作弊绕过策略：
 * 1. Grim 模式 - 发送交互包绕过检测
 * 2. StrafeFix - 修复横向移动
 * 3. AirStrict - 空中模式绕过
 * 4. InventoryMove - 背包中移动
 * 5. Strict 模式 - 严格的发包顺序
 */
class NoSlow : Module(
    MeteorPlusPlusAddon.METEORPLUSPLUS_CATEGORY,
    "NoSlow",
    "防止使用物品时减速，包含 Grim 绕过"
) {
    private val sgGeneral = settings.getDefaultGroup()
    private val sgBlocks = settings.createGroup("Blocks")

    // ==================== 通用设置 ====================
    private val strict = sgGeneral.add(BoolSetting.Builder()
        .name("Strict")
        .description("严格模式（某些反作弊需要）")
        .defaultValue(false)
        .build())

    private val airStrict = sgGeneral.add(BoolSetting.Builder()
        .name("AirStrict")
        .description("空中严格模式")
        .defaultValue(false)
        .build())

    private val grim = sgGeneral.add(BoolSetting.Builder()
        .name("Grim")
        .description("Grim 反作弊绕过")
        .defaultValue(false)
        .build())

    private val grimNew = sgGeneral.add(BoolSetting.Builder()
        .name("GrimV3")
        .description("Grim v3 绕过")
        .defaultValue(false)
        .build())

    private val strafeFix = sgGeneral.add(BoolSetting.Builder()
        .name("StrafeFix")
        .description("修复横向移动")
        .defaultValue(false)
        .build())

    private val inventoryMove = sgGeneral.add(BoolSetting.Builder()
        .name("InventoryMove")
        .description("允许在背包界面移动")
        .defaultValue(true)
        .build())

    private val arrowMove = sgGeneral.add(BoolSetting.Builder()
        .name("ArrowMove")
        .description("允许使用方向键看")
        .defaultValue(false)
        .visible { inventoryMove.get() }
        .build())

    private val items = sgGeneral.add(BoolSetting.Builder()
        .name("Items")
        .description("使用物品时无减速")
        .defaultValue(true)
        .build())

    private val sneak = sgGeneral.add(BoolSetting.Builder()
        .name("Sneak")
        .description("潜行时无减速")
        .defaultValue(false)
        .build())

    private val crawl = sgGeneral.add(BoolSetting.Builder()
        .name("Crawl")
        .description("爬行时无减速")
        .defaultValue(false)
        .build())

    private val shields = sgGeneral.add(BoolSetting.Builder()
        .name("Shields")
        .description("使用盾牌时无减速")
        .defaultValue(true)
        .build())

    // ==================== 方块设置 ====================
    private val webs = sgBlocks.add(BoolSetting.Builder()
        .name("Webs")
        .description("蜘蛛网中无减速")
        .defaultValue(false)
        .build())

    private val berryBush = sgBlocks.add(BoolSetting.Builder()
        .name("BerryBush")
        .description("甜莓灌木丛中无减速")
        .defaultValue(false)
        .build())

    private val webSpeed = sgBlocks.add(DoubleSetting.Builder()
        .name("WebMultiplier")
        .description("蜘蛛网中的速度倍率")
        .defaultValue(1.0)
        .min(0.0)
        .sliderMax(1.0)
        .visible { webs.get() || berryBush.get() }
        .build())

    private val soulSand = sgBlocks.add(BoolSetting.Builder()
        .name("SoulSand")
        .description("灵魂沙上无减速")
        .defaultValue(false)
        .build())

    private val honeyBlock = sgBlocks.add(BoolSetting.Builder()
        .name("HoneyBlock")
        .description("蜂蜜块上无减速")
        .defaultValue(false)
        .build())

    private val slimeBlock = sgBlocks.add(BoolSetting.Builder()
        .name("SlimeBlock")
        .description("粘液块上无减速")
        .defaultValue(false)
        .build())

    // ==================== 状态变量 ====================
    private var sneaking = false

    override fun onDeactivate() {
        if (airStrict.get() && sneaking && mc.player != null) {
            mc.player!!.isSneaking = false
        }
        sneaking = false
    }

    @EventHandler
    private fun onTick(event: TickEvent.Pre) {
        if (mc.player == null || mc.world == null) return

        // Grim 模式 - 发送交互包
        if (grim.get() && mc.player!!.isUsingItem && !mc.player!!.isSneaking && items.get()) {
            if (mc.player!!.activeHand == Hand.OFF_HAND && checkStack(mc.player!!.offHandStack)) {
                sendInteractPacket(Hand.MAIN_HAND)
            } else if (checkStack(mc.player!!.mainHandStack)) {
                sendInteractPacket(Hand.OFF_HAND)
            }
        }

        // AirStrict 模式
        if (airStrict.get() && !mc.player!!.isUsingItem) {
            sneaking = false
            mc.player!!.isSneaking = false
        }

        // InventoryMove
        if (inventoryMove.get() && checkScreen()) {
            // 简化实现：允许在背包中移动
            // 实际应该检查按键状态
        }

        // Grim + Webs
        if ((grim.get() || grimNew.get()) && webs.get()) {
            val bb = if (grim.get()) mc.player!!.boundingBox.expand(1.0) else mc.player!!.boundingBox
            for (pos in getIntersectingWebs(bb)) {
                mc.networkHandler?.sendPacket(
                    net.minecraft.network.packet.c2s.play.PlayerActionC2SPacket(
                        net.minecraft.network.packet.c2s.play.PlayerActionC2SPacket.Action.STOP_DESTROY_BLOCK,
                        pos,
                        net.minecraft.util.math.Direction.UP
                    )
                )
            }
        }

        // AirStrict + 减速检查
        if (airStrict.get() && !sneaking && checkSlowed()) {
            sneaking = true
            mc.player!!.isSneaking = true
        }

        handleMovementSlowdown()
    }

    @EventHandler
    private fun onPacketSend(event: PacketEvent.Send) {
        if (mc.player == null || mc.world == null || mc.isInSingleplayer) return

        val packet = event.packet

        // Strict 模式 - 切换物品
        if (packet is net.minecraft.network.packet.c2s.play.UpdateSelectedSlotC2SPacket) {
            if (strict.get() && checkSlowed()) {
                // 严格模式逻辑
                return
            }
        }

        // Strict 模式 - 移动包
        if (packet is PlayerMoveC2SPacket && strict.get()) {
            if (mc.player!!.isUsingItem) {
                mc.player!!.stopUsingItem()
            }
            // MC 1.21.8: 不能直接发送 RELEASE_SHIFT_KEY，使用其他方式
        }
    }

    private fun handleMovementSlowdown() {
        if (mc.player == null) return

        // Sneak/Crawl 减速
        if ((sneak.get() && mc.player!!.isSneaking) || (crawl.get() && mc.player!!.isSneaking)) {
            val f = 1f / mc.player!!.getMovementSpeed().toFloat()
            mc.player!!.velocity = mc.player!!.velocity.multiply(f.toDouble(), 1.0, f.toDouble())
        }

        // 减速检查
        if (checkSlowed()) {
            mc.player!!.velocity = mc.player!!.velocity.multiply(5.0, 1.0, 5.0)
        }
    }

    private fun checkGrimNew(): Boolean {
        return !(mc.player!!.isSneaking || mc.player!!.isClimbing || mc.player!!.fallDistance >= 5) ||
               (mc.player!!.age > 1 && mc.player!!.age % 2 != 0)
    }

    fun checkSlowed(): Boolean {
        return (!grimNew.get() || checkGrimNew()) &&
               !mc.player!!.isClimbing &&
               !mc.player!!.isSneaking &&
               ((mc.player!!.isUsingItem && items.get()) ||
               (mc.player!!.isBlocking && shields.get() && !grimNew.get() && !grim.get()))
    }

    private fun checkStack(stack: net.minecraft.item.ItemStack): Boolean {
        return !(stack.isOf(Items.CROSSBOW) ||
                stack.isOf(Items.BOW) ||
                stack.isOf(Items.TRIDENT) ||
                stack.isOf(Items.SPYGLASS))
    }

    private fun sendInteractPacket(hand: Hand) {
        mc.networkHandler?.sendPacket(
            PlayerInteractItemC2SPacket(hand, 0, mc.player!!.yaw, mc.player!!.pitch)
        )
    }

    private fun checkScreen(): Boolean {
        return mc.currentScreen == null ||
               mc.currentScreen is net.minecraft.client.gui.screen.ingame.InventoryScreen ||
               mc.currentScreen is net.minecraft.client.gui.screen.ingame.CreativeInventoryScreen ||
               mc.currentScreen is net.minecraft.client.gui.screen.ingame.GenericContainerScreen
    }

    fun getIntersectingWebs(boundingBox: Box): List<net.minecraft.util.math.BlockPos> {
        val blocks = ArrayList<net.minecraft.util.math.BlockPos>()
        val minX = boundingBox.minX.toInt()
        val maxX = boundingBox.maxX.toInt()
        val minY = boundingBox.minY.toInt()
        val maxY = boundingBox.maxY.toInt()
        val minZ = boundingBox.minZ.toInt()
        val maxZ = boundingBox.maxZ.toInt()

        for (x in minX..maxX) {
            for (y in minY..maxY) {
                for (z in minZ..maxZ) {
                    val pos = net.minecraft.util.math.BlockPos(x, y, z)
                    if (mc.world!!.getBlockState(pos).block is CobwebBlock) {
                        blocks.add(pos)
                    }
                }
            }
        }
        return blocks
    }

    fun noSlow(): Boolean {
        return isActive && checkSlowed()
    }

    fun getStrafeFix(): Boolean {
        return isActive && strafeFix.get()
    }
}
